crates/common/src/runtime_client.rs

```diff
@@ -0,0 +1,258 @@
+    Ipc(Ipc),
+}
+
+/// Error type for the runtime provider
+#[derive(Error, Debug)]
+pub enum RuntimeClientError {
+    /// Internal provider error
+    #[error(transparent)]
+    ProviderError(ProviderError),
+
+    /// Failed to lock the client
+    #[error("Failed to lock the client")]
+    LockError,
+
+    /// Invalid URL scheme
+    #[error("URL scheme is not supported: {0}")]
+    BadScheme(String),
+
+    /// Invalid file path
+    #[error("Invalid IPC file path: {0}")]
+    BadPath(String),
+}
+
+impl RpcError for RuntimeClientError {
+    fn as_error_response(&self) -> Option<&JsonRpcError> {
+        match self {
+            RuntimeClientError::ProviderError(err) => err.as_error_response(),
+            _ => None,
+        }
+    }
+
+    fn as_serde_error(&self) -> Option<&serde_json::Error> {
+        match self {
+            RuntimeClientError::ProviderError(e) => e.as_serde_error(),
+            _ => None,
+        }
+    }
+}
+
+impl From<RuntimeClientError> for ProviderError {
+    fn from(src: RuntimeClientError) -> Self {
+        match src {
+            RuntimeClientError::ProviderError(err) => err,
+            _ => ProviderError::JsonRpcClientError(Box::new(src)),
+        }
+    }
+}
+
+/// A provider that connects on first request allowing handling of different provider types at
+/// runtime
+#[derive(Clone, Debug, Error)]
+pub struct RuntimeClient {
+    client: Arc<RwLock<Option<InnerClient>>>,
+    url: Url,
+    max_retry: u32,
+    timeout_retry: u32,
+    initial_backoff: u64,
+    timeout: Duration,
+    /// available CUPS
+    compute_units_per_second: u64,
+    jwt: Option<String>,
+}
+
+impl ::core::fmt::Display for RuntimeClient {
+    fn fmt(&self, f: &mut ::core::fmt::Formatter<'_>) -> ::core::fmt::Result {
+        write!(f, "RuntimeClient")
+    }
+}
+
+fn build_auth(jwt: String) -> eyre::Result<Authorization> {
+    // Decode jwt from hex, then generate claims (iat with current timestamp)
+    let jwt = hex::decode(jwt)?;
+    let secret = JwtKey::from_slice(&jwt).map_err(|err| eyre::eyre!("Invalid JWT: {}", err))?;
+    let auth = JwtAuth::new(secret, None, None);
+    let token = auth.generate_token()?;
+
+    // Essentially unrolled ethers-rs new_with_auth to accomodate the custom timeout
+    let auth = Authorization::Bearer(token);
+
+    Ok(auth)
+}
+
+impl RuntimeClient {
+    /// Creates a new dynamic provider from a URL
+    pub fn new(
+        url: Url,
+        max_retry: u32,
+        timeout_retry: u32,
+        initial_backoff: u64,
+        timeout: Duration,
+        compute_units_per_second: u64,
+        jwt: Option<String>,
+    ) -> Self {
+        Self {
+            client: Arc::new(RwLock::new(None)),
+            url,
+            max_retry,
+            timeout_retry,
+            initial_backoff,
+            timeout,
+            compute_units_per_second,
+            jwt,
+        }
+    }
+
+    async fn connect(&self) -> Result<InnerClient, RuntimeClientError> {
```

Comment:
let's maybe add a bit of docs here, and document the function a bit as it's pretty big